rfcs/0000-compiled-runtime.md

```diff
@@ -0,0 +1,473 @@
+]);
+```
+
+An identity function is used to flag call sites for extraction later. If the style rules are not declared at the top of the module they will be hoisted if safe to the top of the module.
+
+```jsx
+/** @jsxImportSource @compiled/react */
+
+function Comp() {
+  return <div css={css({ color: 'black' })} />;
+}
+
+// 👇 transforms to
+
+const _1 = i(['_syaz11x8', ['._syaz11x8{color:black}']]);
+
+function Comp() {
+  return <div css={_1} />;
+}
+```
+
+When referencing an expression Compiled would try to statically evaluate it, if it fails it would be added as the third item in the array along with all other dynamic styles.
+
+```jsx
+import { N800 } from '@atlaskit/theme/colors';
+
+const styles = css({ color: N800 });
+
+// 👇 transforms to
+
+const styles = i(['_syaz4rde', ['._syaz4rde{color: var(--_kmurgp)}'], [['--_kmurgp', N800]]]);
+```
+
+When composing style rules together in an array they would naturally work **without extra AST transformation**.
+
+```jsx
+const colorStyles = css({
+  color: 'black',
+});
+
+const hoverStyles = css({
+  ':hover': {
+    color: 'red',
+  },
+});
+
+const styles = css([colorStyles, hoverStyles]);
+
+// 👇 transforms to
+
+const colorStyles = i(['_syaz11x8', ['._syaz11x8{color:black}']]);
+
+const hoverStyles = i(['_30l35scu', ['._30l35scu:hover{color:red}']]);
+
+const styles = i([colorStyles, hoverStyles]);
+```
+
+**Considerations**
+
+- Could there be a better way to mark sites for extraction?
+- What would the best data structure be for these styles, what would encourage the fastest HOT path? I've chosen arrays as they don't take much bundle size vs. objects.
+- How can we de-duplicate style declarations in the same module?
+- Composition of style rules via object rest wouldn't be possible without further AST transformation, though perhaps this remains unsupported? Have only a single way to do composition.
+
+### Keyframes
+
+Keyframes are a bit tricky as they can't easily lean on the same pattern the previous API leans on while still encouraging static compiled strings that can be easily extracted later in time, including module traversal intricacies.
+
+This is probably the weakest part of this RFC currently and would love any thoughts of how best to handle them.
+
+```jsx
+import { css, keyframes } from '@compiled/react';
+
+const fadeIn = keyframes({
+  to: { opacity: 1 },
+});
+
+const styles = css({
+  animationName: fadeIn,
+  animationDuration: '2s',
+});
+
+// 👇 transforms to
+
+const styles = i([
+  '_5sagymdr _j7hqa2t1',
+  [
+    '._j7hqa2t1{animation-name:kfwl3rt}',
+    '@keyframes kfwl3rt{to{opacity:1}}',
+    '._5sagymdr{animation-duration:2s}',
+  ],
+]);
+```
+
+Note that it in essence still behaves the same as it does today. The keyframes call site is erased and merged into the style rule. I had some an idea for keyframes to build its own style rule declaration (below) but it broke down when wanting to define animation in pseudo elements/classes.
```

Comment:
Can you elaborate more on:

> but it broke down when wanting to define animation in pseudo elements/classes.

Would be interested in seeing the example in practice, as the one below looks fine to me?

Reply:
It's hypothetical of course but the overarching goal is how can we architect things so it requires the least amount of AST traversal (and knowledge of other nodes). If I decided to apply an animation only on hover:

```
import { css, keyframes } from '@compiled/react';

// Has no knowledge of styles
const fadeIn = keyframes({
  to: { opacity: 1 },
});

// Has no knowledge of fadeIn animation
const styles = css({
  animationDuration: '2s',
  ':hover': {
      animationName: fadeIn,
  }
});
```

How could we make this work when they don't have knowledge of each other? 🤔

Reply:
Right, so I encountered a similar kind of problem in my experimental implementation. They have an inherent dependency on each other, so I don't think there is any easy way around this. I handled the presence of `animationName` specifically (which breaks down for the tagged template expression syntax), and generated:

```js
const _3 = \\"._1pgl1ytf{animation-timing-function:ease-in-out}\\";
const _2 = \\"._j7hq13il{animation-name:var(--_129ax8o)}\\";
const _ = \\"._5sagymdr{animation-duration:2s}\\";

const fadeOut = {
  name: "kug81sc",
  sheet: "@keyframes kug81sc{25%,0%{opacity:1}25%{opacity:0.75}50%{opacity:0.5}to{opacity:0}}",
};

<CC>
  <CS>{[_, _2, _3]}</CS>
  <div
    className={ax([\\"_5sagymdr _j7hq13il _1pgl1ytf\\"])}
    style={{
      "--_129ax8o\\": ix(fadeOut),
    }}
  />
</CC>
```

The above works a bit differently to yours in that `keyframes()` does not generate an `animation-name` property and value, but instead has `name`. I keep the `animationName` in the original css and utilise the style prop to get the animation name at runtime. But now I have the problem where the `CS` component does not contain the keyframe css, since it's this weird global thing. I think in my case I need to include code evaluation to avoid this problem, and can then remove the need for the `style` prop.

Your approach is more self-contained by having `animation-name` in the keyframe, but how are you omitting the original `animationName` from the output `styles` to get `i([fadeIn, ...])`. Are you also looking specifically for the presence of the animation key?

As for hovering specifically with your implementation, you need code evaluation to get the real animation name as `kfwl3rt` or do it dynamically like in my version above.

Reply:
Full disclosure most of this RFC isn't implemented yet 😄. Yeah it would handle animation keys as a special case and do something differently.

Love your use of dynamic styles! That actually makes me wonder if we could just reference things by value and then have it be a special case that needs to be handled during style extraction.

```jsx
const fadeIn = keyframes({
  to: { opacity: 1 },
});

const styles = css({
  animationDuration: '2s',
  ':hover': {
      animationName: fadeIn,
  }
});

// transforms to

const fadeIn = i(['kfwl3rt', ['@keyframes kfwl3rt{to{opacity:1}}']]);

const styles = i(['_j7hqa2t1 _5sagymdr', [`._j7hqa2t1{animation-name:${fadeIn[0]}}`, '._5sagymdr{animation-duration:2s}']]);
```

It makes style extraction harder, especially if cross module boundaries which is something we'd like to eliminate I imagine... hence your point about code eval. 

Tough! Something for us to chew on TBH.

Reply:
Bit of late night thinking in bed hehe so I got up to write it down. I've had a thought where we could keep the separation but still get the result we want using CSS custom properties.

```js
const fadeIn = keyframes({
  to: { opacity: 1 },
});

const styles = css({
  animationDuration: '2s',
  ':hover': {
      animationName: fadeIn,
  }
});

// transforms to

const fadeIn = i(['_5qwe', ['_5qwe{--anim:kfwl3rt}', '@keyframes kfwl3rt{to{opacity:1}}']]);

const styles = i(['_j7hqa2t1 _5sagymdr', [`._j7hqa2t1:hover{animation-name:var(--anim)}}`, '._5sagymdr{animation-duration:2s}']]);
```

**The crux of the solution**
- Any use of animation or animation name will always get transformed to be a CSS variable (probably something hashed, but it should be the same).
- Keyframes generates some styles with the animation name applied to the anim CSS custom property

**Pitfalls**
- It would only work with one animation on an element (as the custom property is static) so it isn't a perfect solution. But maybe there's something here with further digging.